GNUmakefile

```diff
@@ -8,6 +8,10 @@ default: build
 build: fmtcheck
 	go install
 
+dev: fmtcheck
+	go build -o terraform-provider-vault
+	mv terraform-provider-vault ~/.terraform.d/plugins
```

Comment:
This is unrelated to directory generation, just a helpful method for developers.

Reply:
nit: if `~/.terraform.d/plugins` directory doesn't already exist, this moves the compiled binary to that path _as `~/.terraform.d/plugins`. 

<details>

```
[terraform-provider-vault][add-directory-structure](4)$ ls -la ~/.terraform.d
total 16
drwxr-xr-x   4 clint  staff   128 Apr 27 13:59 .
drwxr-xr-x+ 73 clint  staff  2336 Apr 27 13:59 ..
-rw-r--r--   1 clint  staff   319 Apr 23 06:55 checkpoint_cache
-rw-r--r--   1 clint  staff   394 Apr 20 07:49 checkpoint_signature

[terraform-provider-vault][add-directory-structure](4)$ make dev
==> Checking that code complies with gofmt requirements...
go build -o terraform-provider-vault
mv terraform-provider-vault ~/.terraform.d/plugins

[terraform-provider-vault][add-directory-structure](4)$ ls -la ~/.terraform.d
total 72944
drwxr-xr-x   5 clint  staff       160 Apr 27 13:59 .
drwxr-xr-x+ 73 clint  staff      2336 Apr 27 13:59 ..
-rw-r--r--   1 clint  staff       319 Apr 23 06:55 checkpoint_cache
-rw-r--r--   1 clint  staff       394 Apr 20 07:49 checkpoint_signature
-rwxr-xr-x   1 clint  staff  37335636 Apr 27 13:59 plugins

[terraform-provider-vault][add-directory-structure](4)$ ~/.terraform.d/plugins
This binary is a plugin. These are not meant to be executed directly.
Please execute the program that consumes these plugins, which will
load any plugins automatically
```
</details>

I believe either a trailing slash `/` on the path, or using the `-p` option for `mkdir` will resolve that.

Terraform searches the current directory for plugins as a convenience for developing, and recommends ~/.terraform.d/plugins` for third-party plugins. I'm not sure we should encourage placing dev provider builds there.

https://www.terraform.io/docs/extend/how-terraform-works.html#plugin-locations

Reply:
https://github.com/terraform-providers/terraform-provider-vault/pull/738/commits/386b63148cc3be9c6bf78c71f12a71ffddad5835